File: examples/cpp/fetch_observations.cpp
// Fetch iMineralogist observations with cursor pagination and rate-limit
// handling, using libcurl for HTTP and nlohmann/json for parsing.
//
// Build (macOS, Homebrew):
//   brew install nlohmann-json
//   c++ -std=c++17 fetch_observations.cpp -lcurl -I"$(brew --prefix nlohmann-json)/include" -o fetch_observations
//
// Build (Debian/Ubuntu):
//   sudo apt install libcurl4-openssl-dev nlohmann-json3-dev
//   g++ -std=c++17 fetch_observations.cpp -lcurl -o fetch_observations
//
// Usage:
//   export IMINERALOGIST_API_KEY=imin_your_key_here
//   ./fetch_observations quartz          # optional mineral filter
#include <curl/curl.h>
#include <nlohmann/json.hpp>

#include <chrono>
#include <cstdlib>
#include <iostream>
#include <string>
#include <thread>

using json = nlohmann::json;

static const std::string kBaseUrl = "https://imineralogist.org/api/v1";

// libcurl hands the response body to this callback in chunks.
static size_t writeToString(char* data, size_t size, size_t nmemb, void* userdata) {
    static_cast<std::string*>(userdata)->append(data, size * nmemb);
    return size * nmemb;
}

struct HttpResponse {
    long status = 0;
    std::string body;
    long retryAfterSeconds = 60; // from the Retry-After header on 429
};

// One GET request with the API-key header. Exits on transport errors;
// HTTP-level errors are returned for the caller to inspect.
static HttpResponse get(const std::string& url, const std::string& apiKey) {
    CURL* curl = curl_easy_init();
    if (!curl) {
        std::cerr << "curl_easy_init failed\n";
        std::exit(1);
    }

    HttpResponse response;
    curl_slist* headers = curl_slist_append(nullptr, ("X-API-Key: " + apiKey).c_str());
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeToString);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response.body);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);

    CURLcode result = curl_easy_perform(curl);
    if (result != CURLE_OK) {
        std::cerr << "request failed: " << curl_easy_strerror(result) << "\n";
        std::exit(1);
    }
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response.status);

    // Retry-After tells us exactly how long to wait when rate limited.
    curl_header* header = nullptr;
    if (curl_easy_header(curl, "Retry-After", 0, CURLH_HEADER, -1, &header) == CURLHE_OK) {
        response.retryAfterSeconds = std::strtol(header->value, nullptr, 10);
        if (response.retryAfterSeconds <= 0) response.retryAfterSeconds = 60;
    }

    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    return response;
}

// URL-encode one query-parameter value (bbox commas, mineral names with
// spaces, and pagination cursors all need it).
static std::string urlEncode(const std::string& value) {
    char* escaped = curl_easy_escape(nullptr, value.c_str(), static_cast<int>(value.size()));
    std::string encoded(escaped);
    curl_free(escaped);
    return encoded;
}

int main(int argc, char** argv) {
    const char* apiKey = std::getenv("IMINERALOGIST_API_KEY");
    if (!apiKey) {
        std::cerr << "Set IMINERALOGIST_API_KEY first (see README).\n";
        return 1;
    }
    curl_global_init(CURL_GLOBAL_DEFAULT);

    // Any GET /observations parameter can be appended the same way —
    // bbox, q, user, mindat_id, observed_from/observed_to, sort, ...
    // Full reference: ../docs/reference.md
    std::string baseQuery = "limit=100";
    if (argc > 1) baseQuery += "&mineral=" + urlEncode(argv[1]);

    std::string cursor;
    int count = 0;
    const int maxRecords = 100;
    while (count < maxRecords) {
        std::string url = kBaseUrl + "/observations?" + baseQuery;
        // Same filters + cursor from the previous page = next page.
        if (!cursor.empty()) url += "&cursor=" + urlEncode(cursor);

        HttpResponse response = get(url, apiKey);

        if (response.status == 429) { // rate limited — wait and retry the same page
            std::cerr << "rate limited — waiting " << response.retryAfterSeconds << "s\n";
            std::this_thread::sleep_for(std::chrono::seconds(response.retryAfterSeconds));
            continue;
        }
        if (response.status != 200) {
            // All errors share one envelope: {"error": {"code", "message"}}
            json error = json::parse(response.body, nullptr, /*allow_exceptions=*/false);
            if (error.contains("error")) {
                std::cerr << error["error"]["code"].get<std::string>() << ": "
                          << error["error"]["message"].get<std::string>() << "\n";
            } else {
                std::cerr << "HTTP " << response.status << " (no error envelope)\n";
            }
            return 1;
        }

        json page = json::parse(response.body);
        for (const json& obs : page["data"]) {
            // community_consensus is null until the community agrees on a name.
            const json& consensus = obs["community_consensus"];
            std::string name = consensus.is_null() ? "(unidentified)"
                                                   : consensus["name"].get<std::string>();
            // location.privacy is "open" or "obscured"; obscured coordinates
            // are deliberately offset — don't attempt to de-obscure them.
            std::cout << obs["observed_at"].get<std::string>() << "  " << name << "  ("
                      << obs["location"]["privacy"].get<std::string>() << ")  "
                      << obs["web_url"].get<std::string>() << "\n";
            if (++count >= maxRecords) break;
        }

        if (page["next_cursor"].is_null()) break; // last page
        cursor = page["next_cursor"].get<std::string>();
    }

    std::cout << "\n" << count
              << " observations. Data © iMineralogist community contributors.\n";
    curl_global_cleanup();
    return 0;
}

File: examples/go/fetch_observations.go
// Fetch iMineralogist observations with typed structs, cursor pagination,
// and rate-limit handling.
//
// Only the standard library is used — no go.mod needed:
//
//	export IMINERALOGIST_API_KEY=imin_your_key_here
//	go run fetch_observations.go -mineral quartz -max 150
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"time"
)

const baseURL = "https://imineralogist.org/api/v1"

// Only the fields this example uses are declared. The API may add new
// response fields at any time; encoding/json ignores unknown fields, which
// is exactly what the versioning policy asks clients to do.
type Observation struct {
	ID         string `json:"id"`
	Title      string `json:"title"`
	ObservedAt string `json:"observed_at"`
	Location   struct {
		Latitude   float64  `json:"latitude"`
		Longitude  float64  `json:"longitude"`
		PrecisionM *float64 `json:"precision_m"` // pointer: may be null
		Privacy    string   `json:"privacy"`     // "open" or "obscured"
	} `json:"location"`
	// Pointer because an observation may have no consensus yet (null).
	CommunityConsensus *struct {
		Name     string `json:"name"`
		MindatID *int   `json:"mindat_id"`
	} `json:"community_consensus"`
	Observer struct {
		Nickname string `json:"nickname"`
	} `json:"observer"`
	WebURL string `json:"web_url"`
}

type listResponse struct {
	Data       []Observation `json:"data"`
	NextCursor *string       `json:"next_cursor"` // null on the last page
}

type apiError struct {
	Error struct {
		Code    string `json:"code"`
		Message string `json:"message"`
	} `json:"error"`
}

func main() {
	mineral := flag.String("mineral", "", "substring filter on mineral names")
	bbox := flag.String("bbox", "", "west,south,east,north in WGS84 degrees")
	maxRecords := flag.Int("max", 100, "stop after this many observations")
	flag.Parse()

	apiKey := os.Getenv("IMINERALOGIST_API_KEY")
	if apiKey == "" {
		fmt.Fprintln(os.Stderr, "Set IMINERALOGIST_API_KEY first (see README).")
		os.Exit(1)
	}

	client := &http.Client{Timeout: 30 * time.Second}

	// Any GET /observations parameter fits here; all filters combine with AND.
	params := url.Values{"limit": {"100"}}
	if *mineral != "" {
		params.Set("mineral", *mineral)
	}
	if *bbox != "" {
		params.Set("bbox", *bbox)
	}

	count := 0
	for {
		req, err := http.NewRequest("GET", baseURL+"/observations?"+params.Encode(), nil)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		req.Header.Set("X-API-Key", apiKey)

		resp, err := client.Do(req)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		if resp.StatusCode == http.StatusTooManyRequests {
			// Rate limited — Retry-After says how many seconds to wait.
			wait, _ := strconv.Atoi(resp.Header.Get("Retry-After"))
			if wait <= 0 {
				wait = 60
			}
			resp.Body.Close()
			fmt.Fprintf(os.Stderr, "rate limited — waiting %ds\n", wait)
			time.Sleep(time.Duration(wait) * time.Second)
			continue // retry the same page
		}
		if resp.StatusCode != http.StatusOK {
			// All errors share one envelope: {"error": {"code", "message"}}
			var apiErr apiError
			if json.NewDecoder(resp.Body).Decode(&apiErr) == nil && apiErr.Error.Code != "" {
				fmt.Fprintf(os.Stderr, "%s: %s\n", apiErr.Error.Code, apiErr.Error.Message)
			} else {
				fmt.Fprintf(os.Stderr, "HTTP %d\n", resp.StatusCode)
			}
			resp.Body.Close()
			os.Exit(1)
		}

		var page listResponse
		err = json.NewDecoder(resp.Body).Decode(&page)
		resp.Body.Close()
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		for _, obs := range page.Data {
			name := "(unidentified)"
			if obs.CommunityConsensus != nil {
				name = obs.CommunityConsensus.Name
			}
			// Obscured coordinates are deliberately offset — precision_m is
			// the obscuring radius. Don't attempt to de-obscure locations.
			fmt.Printf("%s  %-24s %-8s %s\n", obs.ObservedAt, name, obs.Location.Privacy, obs.WebURL)
			count++
			if count >= *maxRecords {
				fmt.Printf("\n%d observations. Data © iMineralogist community contributors.\n", count)
				return
			}
		}

		if page.NextCursor == nil {
			break // last page
		}
		// Same filters + cursor from the previous page = next page.
		params.Set("cursor", *page.NextCursor)
	}
	fmt.Printf("\n%d observations. Data © iMineralogist community contributors.\n", count)
}
